hack/update-snapshots.sh

```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+# hacky script to update snapshots, used by pre-commit
+
+set -e
+
+source hack/sh/rc.sh
+source hack/sh/_functions.sh
+source hack/sh/deps/helm.sh
+
+set -x
+
+# grab all charts that where modified
+declare -a charts
+for file in $@; do
```

Comment:
this needs more explanation for me. What arguments are passed to this script exactly?

Reply:
pre-commit passes all the changed files to it's hook scripts.

I put some more text into the header of the script to further clarify what `$@` is any what it is being used for.

Reply:
why the `| cut -d/ -f1-2`

Reply:
`$@` contains a list of all the changed files but we are only interesed in the chart dir:

`charts/foo-bar/Chart.yaml` => `charts/foo-bar`

Reply:
there is but it won't work for `charts/foo-bar/templates/baz.yaml`

Reply:
Then we would need also a `sort -u` somewhere? aren't we ending up with a list of duplicate charts dir? 

`files=charts/foo-bar/Chart.yaml charts/fooo-bar/templates/baz.yaml` -> `charts=charts/foo-bar charts/foo-bar` ?

Reply:
using an array takes care of this in a convolute bashy way:

```
$ ./hack/update-snapshots.sh charts/foo-bar/Chart.yaml charts/fooo-bar/templates/baz.yaml
+ declare -a charts
+ for file in '$@'
+ charts+=(`echo $file | cut -d/ -f 1-2`)
++ echo charts/foo-bar/Chart.yaml
++ cut -d/ -f 1-2
+ for file in '$@'
+ charts+=(`echo $file | cut -d/ -f 1-2`)
++ echo charts/fooo-bar/templates/baz.yaml
++ cut -d/ -f 1-2
+ for chart in '$charts'
+ helm dep build charts/foo-bar
Error: could not find charts/foo-bar: stat charts/foo-bar: no such file or directory
+ helm unittest --helm3 --update-snapshot charts/foo-bar

### Error:  stat charts/foo-bar: no such file or directory


Charts:      1 failed, 1 errored, 0 passed, 1 total
Test Suites: 0 passed, 0 total
Tests:       0 passed, 0 total
Snapshot:    0 passed, 0 total
Time:        44.247µs

Error: plugin "unittest" exited with error
```

Reply:
Have you tried running it with multiple charts dir? This does not seem to work:

```
$ ~ declare -a arr
$ ~ arr+=("1")
$ ~ arr+=("2")
$ ~ echo $arr
1
$ ~ for i in $arr; do echo $i ; done
1
$ ~ for i in ${arr[@]}; do echo $i ; done
1
2
```

And it needs a `sort -u `  I think.

Tested with Bash version `5.1.4`

[reference](https://linuxhint.com/bash_append_array/)

Reply:
> using an array takes care of this in a convoluted bashy way:
> 
> ```
> $ ./hack/update-snapshots.sh charts/infra-apps/Chart.yaml charts/infra-apps/templates/ingress.yaml charts/infra-apps/Chart.lock 
> + declare -a charts
> + for file in '$@'
> + charts+=(`echo $file | cut -d/ -f 1-2`)
> ++ echo charts/infra-apps/Chart.yaml
> ++ cut -d/ -f 1-2
> + for file in '$@'
> + charts+=(`echo $file | cut -d/ -f 1-2`)
> ++ echo charts/infra-apps/templates/ingress.yaml
> ++ cut -d/ -f 1-2
> + for file in '$@'
> + charts+=(`echo $file | cut -d/ -f 1-2`)
> ++ echo charts/infra-apps/Chart.lock
> ++ cut -d/ -f 1-2
> + for chart in '$charts'
> + helm dep build charts/infra-apps
> walk.go:74: found symbolic link in path: /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/infra-apps/charts/argoconfig resolves to /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/argoconfig
> walk.go:74: found symbolic link in path: /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/infra-apps/charts/argoconfig/charts/common resolves to /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/common
> Hang tight while we grab the latest from your chart repositories...
> ...Successfully got an update from the "sentry" chart repository
> ...Successfully got an update from the "adfinis" chart repository
> ...Successfully got an update from the "argo" chart repository
> ...Successfully got an update from the "incubator" chart repository
> ...Successfully got an update from the "stable" chart repository
> Update Complete. ⎈Happy Helming!⎈
> Saving 1 charts
> Downloading argoconfig from repo https://charts.adfinis.com
> Deleting outdated charts
> + helm unittest --helm3 --update-snapshot charts/infra-apps
> 2021/03/31 10:41:10 found symbolic link in path: /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/infra-apps/charts/argoconfig resolves to /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/argoconfig
> 2021/03/31 10:41:10 found symbolic link in path: /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/infra-apps/charts/argoconfig/charts/common resolves to /Users/hairmare/Documents/git.repos/adfinis-sygroup/helm-charts/charts/common
> 
> ### Chart [ infra-apps ] charts/infra-apps
> 
>  PASS  snapshot tests	charts/infra-apps/tests/certmgr_test.yaml
>  PASS  snapshot tests	charts/infra-apps/tests/ingress_test.yaml
>  PASS  snapshot tests	charts/infra-apps/tests/argocd_test.yaml
> 
> Charts:      1 passed, 1 total
> Test Suites: 3 passed, 3 total
> Tests:       3 passed, 3 total
> Snapshot:    3 passed, 3 total
> Time:        36.458234ms
> ```

You are just reading the first item of the array, this is why you think there is no need for `uniq`. Try echoing `${charts[@]}`

Reply:
hmmm

Reply:
Bash arrays. I hate them.

Reply:
why did I even think writing this in bash is a good idea 🙄 

it works as intended now that I've introduced some `uniq`:

```
$ ./hack/update-snapshots.sh charts/infra-apps/Chart.yaml charts/security-apps/Chart.yaml charts/infra-apps/Chart.lock
+ declare -a charts
+ for file in '$@'
+ charts+=(`echo $file | cut -d/ -f 1-2`)
++ echo charts/infra-apps/Chart.yaml
++ cut -d/ -f 1-2
+ for file in '$@'
+ charts+=(`echo $file | cut -d/ -f 1-2`)
++ echo charts/security-apps/Chart.yaml
++ cut -d/ -f 1-2
+ for file in '$@'
+ charts+=(`echo $file | cut -d/ -f 1-2`)
++ echo charts/infra-apps/Chart.lock
++ cut -d/ -f 1-2
++ echo charts/infra-apps charts/security-apps charts/infra-apps
++ tr ' ' '\n'
++ sort
++ uniq
+ charts='charts/infra-apps
charts/security-apps'
+ for chart in '$charts'
+ echo helm dep build charts/infra-apps
helm dep build charts/infra-apps
+ echo helm unittest --helm3 --update-snapshot charts/infra-apps
helm unittest --helm3 --update-snapshot charts/infra-apps
+ for chart in '$charts'
+ echo helm dep build charts/security-apps
helm dep build charts/security-apps
+ echo helm unittest --helm3 --update-snapshot charts/security-apps
helm unittest --helm3 --update-snapshot charts/security-apps
```